Objects/longobject.c

```diff
@@ -4503,17 +4509,60 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
     }
 
     if (Py_SIZE(a) < 0) {
-        /* Right shifting negative numbers is harder */
-        PyLongObject *a1, *a2;
-        a1 = (PyLongObject *) long_invert(a);
-        if (a1 == NULL)
-            return NULL;
-        a2 = (PyLongObject *) long_rshift1(a1, wordshift, remshift);
-        Py_DECREF(a1);
-        if (a2 == NULL)
+        /*
+            Right shifting negative numbers is harder. For a positive
+            integer a and nonnegative shift, we have:
+
+                (-a) >> shift == -((a + 2**shift - 1) >> shift).
+
+            With shift == wordshift*SHIFT + remshift, 0 <= remshift <= SHIFT,
+            (a + 2**shift - 1) >> shift is equal to
+
+                (a + 2**shift - 1) >> wordshift*SHIFT >> remshift.
+
+            If the bottom `wordshift` digits of a are all zero, this is the
+            equal to
+
+                ((a >> wordshift*SHIFT) + 2**remshift - 1) >> remshift.
+
+            Otherwise, it's equal to:
+
+                ((a >> wordshift*SHIFT) + 2**remshift) >> remshift
+        */
+
+        /* It's convenient for remshift to be positive below. Note that
+           we dealt with the case remshift == wordshift == 0 earlier. */
+        if (remshift == 0) {
+            remshift = PyLong_SHIFT;
+            wordshift -= 1;
+        }
+        assert(wordshift >= 0);
+
+        newsize = -Py_SIZE(a) - wordshift;
+        if (newsize <= 0) {
+            return PyLong_FromLong(-1);
+        }
+        hishift = PyLong_SHIFT - remshift;
+        z = _PyLong_New(newsize);
+        if (z == NULL) {
             return NULL;
-        z = (PyLongObject *) long_invert(a2);
-        Py_DECREF(a2);
+        }
+        Py_SET_SIZE(z, -newsize);
+
+        /* sticky is used to determine whether all dropped words are zero. */
+        sticky = 0;
+        for (j = 0; j < wordshift; j++) {
+            sticky |= a->ob_digit[j];
+        }
+        accum = a->ob_digit[j++] + ((digit)1U << remshift) - (sticky == 0);
+        accum >>= remshift;
+        for (i = 0; j < -Py_SIZE(a); i++, j++) {
+            accum += (twodigits)a->ob_digit[j] << hishift;
```

Comment:
Note that we use `accum |= ...` rather than `accum += ...` in the `Py_SIZE(a) >= 0` branch, but we can't do that here, because there's a possibility of overflow being propagated up the digits.

Reply:
Is there difference between `|=` and `+=` for positive case? If not, we could save several lines of code shared between branches.

Reply:
Both would be correct for the positive case. I'll do some timings to see if there's any detectable performance difference. (I'd be surprised if there were, but you never know.)

Reply:
As expected, no consistently detectable performance difference.